File: beam-native-mobile/NativeSources/iOS/BeamableNotifications/core/Sources/BeamableNotifications/ActionButtons.swift
import Foundation
import UserNotifications

/// Parsing, sanitizing and category-naming for the `buttons` wire key.
///
/// The button model itself lives in `ActionButton.swift` (Foundation only) because the Widget
/// extension needs it; everything here touches `UserNotifications` and stays app/NSE-side.
public enum BeamActionButtons {

    /// How many buttons survive into a synthesized category.
    ///
    /// iOS itself renders up to four actions on an expanded notification, but the console authors
    /// EXACTLY two (`MAX_LIVE_ACTIVITY_BUTTONS`) because the Dynamic Island expanded region only shows
    /// two. Both iOS surfaces must show the same buttons, so the smaller number wins here.
    public static let maxButtons = 2

    /// The `userInfo` key the rail writes (`PushRailService.WriteIntentData` → APNs payload root).
    public static let userInfoKey = "buttons"

    /// Prefix of every synthesized category id. `beam_actions` itself stays the built-in fallback.
    public static let synthesizedPrefix = "beam_actions_"

    /// Category id the SDK registers at init for the built-in `actions` style (Open / Dismiss), used
    /// as the last resort when a payload carries no usable `buttons`. Declared here rather than on
    /// `StyleServicePlugin` because the plugin lives in the NSE target, which core cannot see.
    public static let builtInActionsCategory = "beam_actions"

    /// True for a category id this SDK synthesized (used to scope LRU eviction to our own pool).
    public static func isSynthesized(_ categoryId: String) -> Bool {
        categoryId.hasPrefix(synthesizedPrefix)
    }

    /// Read `buttons` out of a notification's `userInfo`, sanitized and capped.
    public static func parse(userInfo: [AnyHashable: Any]) -> [BeamActionButton] {
        sanitize(parse(userInfo[userInfoKey]))
    }

    /// Tolerant parse of a single `buttons` value.
    ///
    /// The rail stringifies every non-scalar (`extra` is a `Dictionary<string,string>`, and FCM data
    /// values must be strings), so over the wire this is a JSON **string**. A locally-scheduled
    /// notification or a hand-written `simctl push` payload can carry a real array instead — accept
    /// both, exactly like `JSONValue.bmnDecodeStringified` does for `offers` / `campaignData`.
    /// Anything else, or malformed JSON, yields `[]` so the caller falls back rather than failing.
    public static func parse(_ raw: Any?) -> [BeamActionButton] {
        guard let raw = raw else { return [] }

        if let string = raw as? String {
            let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty, let data = trimmed.data(using: .utf8) else { return [] }
            return (try? JSON.decoder.decode([BeamActionButton].self, from: data)) ?? []
        }

        if let array = raw as? [Any] {
            guard JSONSerialization.isValidJSONObject(["b": array]),
                  let data = try? JSONSerialization.data(withJSONObject: array) else { return [] }
            return (try? JSON.decoder.decode([BeamActionButton].self, from: data)) ?? []
        }

        return []
    }

    /// Drop unusable entries, dedupe, and cap.
    ///
    /// A blank `id` can't be routed on tap and a blank `title` renders as an invisible button, so both
    /// are dropped. `com.apple.` ids are rejected because `UNNotificationDefaultActionIdentifier` /
    /// `UNNotificationDismissActionIdentifier` live in that namespace — a collision would make a real
    /// button indistinguishable from "the player tapped the body" in `NotificationManager`'s tap path.
    public static func sanitize(_ buttons: [BeamActionButton]) -> [BeamActionButton] {
        var seen = Set<String>()
        var result: [BeamActionButton] = []
        for button in buttons {
            let id = button.id.trimmingCharacters(in: .whitespacesAndNewlines)
            let title = button.title.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !id.isEmpty, !title.isEmpty else { continue }
            guard !id.hasPrefix("com.apple.") else { continue }
            guard seen.insert(id).inserted else { continue }
            result.append(BeamActionButton(id: id, title: title, role: button.role))
            if result.count == maxButtons { break }
        }
        return result
    }

    /// Deterministic category id for a button set.
    ///
    /// Deterministic matters twice over: the NSE and the app are separate processes that must agree on
    /// the id, and two pushes carrying identical buttons must reuse one registration instead of
    /// leaking a new category per delivery. That rules out Swift's `hashValue`, whose seed is
    /// randomized per process — hence an explicit FNV-1a/64 over a canonical rendering.
    public static func categoryId(for buttons: [BeamActionButton]) -> String {
        guard !buttons.isEmpty else { return builtInActionsCategory }
        let canonical = buttons
            .map { "\($0.id)\u{01}\($0.title)\u{01}\($0.role.lowercased())" }
            .joined(separator: "\u{02}")
        return synthesizedPrefix + fnv1a64Hex(canonical)
    }

    /// FNV-1a, 64-bit, lower-case hex. Stable across processes, launches and OS versions.
    static func fnv1a64Hex(_ string: String) -> String {
        var hash: UInt64 = 0xcbf2_9ce4_8422_2325
        for byte in Array(string.utf8) {
            hash ^= UInt64(byte)
            hash = hash &* 0x0000_0100_0000_01B3
        }
        return String(format: "%016llx", hash)
    }
}

public extension CategorySpec {
    /// Build a category from payload-authored buttons.
    ///
    /// Option mapping mirrors the built-in `beam_actions` pair so a default payload behaves exactly as
    /// before: a `destructive` button is destructive and NOT foreground (like `dismiss`), every other
    /// button is `.foreground` (like `open`) because its tap has to route the deep link, which needs
    /// the app frontmost.
    init(synthesizedFrom buttons: [BeamActionButton]) {
        self.init(
            id: BeamActionButtons.categoryId(for: buttons),
            actions: buttons.map { button in
                ActionSpec(id: button.id,
                           title: button.title,
                           foreground: button.isDestructive ? nil : true,
                           destructive: button.isDestructive ? true : nil,
                           authenticationRequired: nil)
            },
            hiddenPreviewsBodyPlaceholder: nil
        )
    }
}

File: beam-native-mobile/NativeSources/Android/BeamableNotifications/notifications/src/main/java/com/beamable/push/ActionButtons.kt
package com.beamable.push

import org.json.JSONArray

/**
 * Payload-authored action buttons — the Android counterpart of the iOS NSE's `buttons` handling.
 *
 * The console's `actions` style sends the buttons it authored in the push itself, as `buttons`: a JSON
 * array of `{id, title, role}` (a JSON *string* on the wire, since the FCM data map is string→string).
 * Rendering them here is what lets a campaign author's own labels ("Claim" / "No thanks") reach the
 * device, instead of whatever button set the app happened to register under a category id.
 *
 * Android needs none of iOS's category machinery for this: [NotificationBuilder] constructs the
 * notification in-process at delivery time, so a button title is just a string passed to
 * `addAction`. There is no pre-registration and no race.
 *
 * Registered categories still win (see [NotificationBuilder.applyActions]) — an app that registered
 * its own set keeps its behavior unchanged.
 */
internal object ActionButtons {

    /** The `buttons` key as written by `PushRailService.WriteIntentData`. */
    const val KEY_BUTTONS = "buttons"

    /**
     * Built-in category id backing the `actions` style, mirroring iOS's `beam_actions`. An app may
     * register its own category under this id to override the built-in pair.
     */
    const val BUILT_IN_ACTIONS_CATEGORY = "beam_actions"

    /**
     * Buttons kept from a payload. The console authors exactly two (its Live Activity surface renders
     * only two), and both platforms must show the same pair, so the smaller number wins here even
     * though a notification could carry more.
     */
    const val MAX_BUTTONS = 2

    /**
     * Last-resort buttons for `style: "actions"` when the payload carries none and no category is
     * registered. Same labels and ids as the iOS built-in pair, so the two platforms read alike.
     */
    fun builtInActions(): List<NotificationActionSpec> = listOf(
        NotificationActionSpec(id = "open", title = "Open", foreground = true, destructive = false),
        NotificationActionSpec(id = "dismiss", title = "Dismiss", foreground = false, destructive = true),
    )

    /**
     * Parse the `buttons` wire value into action specs, sanitized and capped.
     *
     * Tolerant by design — a malformed value must cost the buttons, never the notification: anything
     * unparseable yields an empty list and the caller falls back. Entries with a blank `id` (nothing to
     * dispatch on tap) or blank `title` (an invisible button) are dropped, duplicate ids collapse
     * first-wins, and `role: "destructive"` maps onto [NotificationActionSpec.destructive].
     */
    fun parse(raw: String?): List<NotificationActionSpec> {
        if (raw.isNullOrBlank()) return emptyList()
        val array = try {
            JSONArray(raw.trim())
        } catch (_: Exception) {
            return emptyList()
        }

        val result = ArrayList<NotificationActionSpec>(MAX_BUTTONS)
        val seen = HashSet<String>()
        for (i in 0 until array.length()) {
            val obj = array.optJSONObject(i) ?: continue
            val id = obj.optString("id").trim()
            val title = obj.optString("title").trim()
            if (id.isEmpty() || title.isEmpty()) continue
            if (!seen.add(id)) continue
            val destructive = obj.optString("role").trim().equals("destructive", ignoreCase = true)
            result.add(
                NotificationActionSpec(
                    id = id,
                    title = title,
                    // Mirrors the iOS mapping: a destructive button is not a foreground open, every
                    // other button is (its tap has to route the deep link, which needs the app up).
                    foreground = !destructive,
                    destructive = destructive,
                )
            )
            if (result.size == MAX_BUTTONS) break
        }
        return result
    }
}
